src/linear_algebra/lagrange.lean

```diff
@@ -1,224 +1,540 @@
+and `0` at `v j` for `i ≠ j`.
+* `lagrange.interpolate v r` where `r : ι → F` is a function from the fintype to the field: the
+Lagrange interpolant that evaluates to `r i` at `x i` for all `i : ι`. The `r i` are the _values_
+associated with the _nodes_`x i`.
+* `lagrange.interpolate_at v f`, where `v : ι ↪ F` and `ι` is a fintype, and `f : F → F` is a
+function from the field to itself: this is the Lagrange interpolant that evaluates to `f (x i)`
+at `x i`, and so approximates the function `f`. This is just a special case of the general
+interpolation, where the values are given by a known function `f`.
+-/
 
-* `lagrange.basis s x` where `s : finset F` and `x : F`: the Lagrange basis polynomial
-  that evaluates to `1` at `x` and `0` at other elements of `s`.
-* `lagrange.interpolate s f` where `s : finset F` and `f : F → F`: the Lagrange interpolant
-  that evaluates to `f x` at `x` for `x ∈ s`.
+/-
+These theorems are independent of the theory of Lagrange interpolants, though they are highly
+related to them. They are the theorems that guarantee that a polynomial of bounded degree, when
+specified on sufficient points, is completely determined. Both `fintype` and `finset` versions are
+provided.
 -/
 
-noncomputable theory
-open_locale big_operators classical polynomial
+open_locale polynomial big_operators
+open polynomial
+
+section polynomial_determination
+
+variables {R : Type*} [comm_ring R] [is_domain R] {f g : R[X]}
+
+section finset
+open function fintype
+variables (s : finset R)
+
+theorem eq_zero_of_degree_lt_of_eval_eq_zero (degree_f_lt : f.degree < s.card)
+  (eval_f : ∀ x ∈ s, f.eval x = 0) : f = 0 :=
+begin
+  rw ← mem_degree_lt at degree_f_lt, simp_rw eval_eq_sum_degree_lt_equiv degree_f_lt at eval_f,
+  rw ← degree_lt_equiv_eq_zero_iff_eq_zero degree_f_lt,
+  exact matrix.eq_zero_of_forall_index_sum_mul_pow_eq_zero
+  (injective.comp (embedding.subtype _).inj' (equiv_fin_of_card_eq (card_coe _)).symm.injective)
+  (λ _, eval_f _ (finset.coe_mem _))
+end
+
+theorem eq_of_degree_sub_lt_of_eval_eq (degree_fg_lt : (f - g).degree < s.card)
+  (eval_fg : ∀ x ∈ s, f.eval x = g.eval x) : f = g :=
+begin
+  rw ← sub_eq_zero, refine eq_zero_of_degree_lt_of_eval_eq_zero _ degree_fg_lt _,
+  simp_rw [eval_sub, sub_eq_zero], exact eval_fg
+end
+
+theorem eq_of_degrees_lt_of_eval_eq (degree_f_lt : f.degree < s.card)
+  (degree_g_lt : g.degree < s.card) (eval_fg : ∀ x ∈ s, f.eval x = g.eval x) : f = g :=
+begin
+  rw ← mem_degree_lt at degree_f_lt degree_g_lt,
+  refine eq_of_degree_sub_lt_of_eval_eq _ _ eval_fg,
+  rw ← mem_degree_lt, exact submodule.sub_mem _ degree_f_lt degree_g_lt
+end
+end finset
+
+section indexed
+
+variables {ι : Type*} {v : ι → R} (s : finset ι)
+
+theorem eq_zero_of_degree_lt_of_eval_index_eq_zero (hvs : set.inj_on v s)
+  (degree_f_lt : f.degree < s.card) (eval_f : ∀ i ∈ s, f.eval (v i) = 0) : f = 0 :=
+begin
+  classical, rw ← finset.card_image_of_inj_on hvs at degree_f_lt,
```

Comment:
```suggestion
  classical, 
  rw ← finset.card_image_of_inj_on hvs at degree_f_lt,
```
There's something in the style guide about not putting multiple tactic invocations on one line unless you are closing a proof in one line.